metagpt/actions/design.py

```diff
@@ -0,0 +1,130 @@
+# -*- coding: utf-8 -*-
+# @Date    : 2023/8/17 13:43
+# @Author  : stellahong (stellahong@fuzhi.ai)
+# @Desc    :
+
+# Standard library imports
+from functools import wraps
+from typing import Callable, Any, List, Optional, Tuple
+
+# Local library imports
+from metagpt.actions import Action
+from metagpt.logs import logger
+from metagpt.utils.common import OutputParser
+from metagpt.prompts.sd_design import (
+    SD_PROMPT_KW_OPTIMIZE_TEMPLATE,
+    SD_PROMPT_IMPROVE_OPTIMIZE_TEMPLATE,
+    FORMAT_INSTRUCTIONS,
+    PROMPT_OUTPUT_MAPPING
+)
+from metagpt.utils.resp_parse import flatten_json_structure, try_parse_json
+
+# A default template for the system primer.
+SYSTEM_PRIMER_TEMPLATE = "Act like you are a terminal and always format your response as json. Always return exactly {answer_count} answers per question in English."
+
+
+class Tool:
+    """Define a tool with its name, function and description."""
+    
+    def __init__(self, name: str, func: Callable, description: str) -> None:
+        """Initialize tool."""
+        self.name = name
+        self.func = func
+        self.description = description
+
+
+# Decorator for the BaseModelAction to wrap it with system primer details
+def system_primer_decorator(func):
+    @wraps(func)
+    async def wrapper(*args, **kwargs):
+        system_primer = SYSTEM_PRIMER_TEMPLATE.format(answer_count=kwargs.get('answer_count', 1))
+        logger.info(system_primer)
+        return await func(*args, system_primer=system_primer, **kwargs)
+    
+    return wrapper
+
+
+class BaseModelAction(Action):
+    
+    def __init__(self, name: str = "", description: str = "", *args, **kwargs):
+        super().__init__(name, *args, **kwargs)
+        self.desc = description
+    
+    async def handle_response(self, resp: str) -> Any:
+        """Handle JSON response and extract value."""
+        try:
+            resp_json = flatten_json_structure(try_parse_json(resp))
+            logger.info(resp_json)
+            return resp_json
+        
+        except Exception:
+            logger.error(f" JSON response {resp_json}!")
```

Comment:
In exceptional situations, the resp_json variable might not have been defined. This could lead to a NameError exception.

Reply:
ok